packages/web/src/repositories/leaderboard/leaderboard-repository-impl.ts

```diff
@@ -0,0 +1,96 @@
+import { NetworkClient } from "@common/clients/network-client";
+import { CommonError } from "@common/errors";
+import { LeaderboardRepository } from "./leaderboard-repository";
+import { GetLeaderboardRequest, UpdateLeaderboardHiddenStateRequest } from "./request";
+import {
+  GetLeaderboardResponse,
+  GetLeaderboardByAddressResponse,
+  nullLeaderboardInfo,
+  UpdateLeaderboardHiddenStateResponse,
+} from "./response";
+
+export class LeaderboardRepositoryImpl implements LeaderboardRepository {
+  private networkClient: NetworkClient | null;
+
+  constructor(networkClient: NetworkClient | null) {
+    this.networkClient = networkClient;
+  }
```

Comment:
Since the constructor accepts a NetworkClient | null type, networkClient can be null. However, the current methods in the class check for if (!this.networkClient), which suggests that a client is required for all operations.

```suggestion
  constructor(private readonly networkClient: NetworkClient) {}
```

Therefore, if the constructor requires a non-null client, it seems possible to remove all these checks.

Reply:
I've reviewed @notJoon's suggestion, but I think it's better to keep the current approach (allowing NetworkClient | null), for the following reasons

Maintaining the flexibility of the dependency injection pattern
‘Since we're using the pattern of injecting clients, I think it's better to keep it the way it is.’

1. The value of explicit type declarations: By explicitly declaring the type NetworkClient | null, we make it clear to anyone reading the code that this class can be initialised without a client.

2. Codebase consistency: It's likely that other repository classes currently follow a similar pattern. Changing just one class can break the consistency of your codebase.

3. Flexible instance creation: Allowing nulls gives you the flexibility to respond to different scenarios (lazy initialisation, conditional client setup, etc.).

4. Ease of testing: In a test environment, it's easy to create and test instances without real network clients.

Reply:
Here's my thoughts:

* **Maintaining Flexibility in Dependency Injection**: While allowing `null` in dependency injection patterns may offer more flexibility, this approach can undermine the intent of dependency injection. One of the core principles of dependency injection is managing component coupling through clear contracts. For essential dependencies, requiring non-null values in the constructor better aligns with this principle. [^1]

* **Value of Explicit Type Declarations**: Explicitly declaring types as `NetworkClient | null` may help with code readability, but if you need to perform `null` checks in every method, this implies that the dependency should never actually be `null`. In this case, filtering to only accept the `NetworkClient` type could be considered more explicit.

* **Codebase Consistency**: If other classes in the codebase follow similar patterns, maintaining consistency is important (as doing things differently could cause problems). However, I believe gradually improving the entire codebase with better patterns would be beneficial in the long term.

* **Flexible Instance Creation**: While allowing `null` provides flexibility in scenarios like lazy initialization or conditional client setup, these cases can be adequately addressed through other approaches. Of course, this point is somewhat ambiguous since implementing such changes might require significant effort.

* **Testability**: I understand that in test environments, it's easier to create and test instances without actual network clients. However, using mocks or stubs to replace real dependencies is a common testing approach.

I fully understand the reasoning for maintaining the current code structure. However, in the long term, I believe improving the code to provide clear contracts and type safety would be better. i'll approve this PR. But please consider this in the future work. Thank you.

[^1]: https://en.wikipedia.org/wiki/Dependency_inversion_principle

Reply:
thank you for your detailed review. Your deep understanding of the dependency injection pattern and your thoughts on improving code quality were very helpful.

As you say, I think incrementally improving the codebase is a better way to go in the long run

For now, we'll keep the PR as is, but we'll prioritise these improvements in future sprints. Thanks again, for your insightful suggestions on improving code quality.